### PR TITLE
act: remove shaded fork join pool implementation

### LINKED ISSUE #26180: Remove embedded FJP
Use standard JDK FJP, performance investigation needed.

Previous attempt https://github.com/akka/akka/pull/22560

### PR DESCRIPTION
**Pull request by jrudolph**
I ran a series of benchmarks and there seem to be different behaviors between the existing and JDK ForkJoinPools especially on JDK11 and Graal. The results have been really inconclusive about which one would be a benefit overall.

In lack of better evidence, there is not much reason to keep our own copy, so there it goes.

(I couldn't reproduce the high system CPU usage with more recent JDKs that @patriknw noticed when we had it on the table the last time in #22560.)

Fixes #26180.

### CODE REVIEW

**Comment by patriknw** [comment]
I tried https://github.com/akka/akka/pull/26834 with JDK 8 and 11 on my mbp and the results are looking very good with JDK 11.

Tested with
```
jmh:run -i 7 -wi 3 -f2 -t1 akka.actor.ForkJoinActorBenchmark
```

With JDK 8 it improved from 22 to 42 M msg/s.
With JDK 11 it improved from 19 to 62 M msg/s.

I was positively surprised, so re-ran the JDK 11 test and it had same result (64 M msg/s) again.

```
jdk 8, jdk fjp
[info] ForkJoinActorBenchmark.pingPongLessActorsThanCores  akka.dispatch.SingleConsumerOnlyUnboundedMailbox         12     50  thrpt   14  12319902.286 ± 1258285.806  ops/s
[info] ForkJoinActorBenchmark.pingPongMoreActorsThanCores  akka.dispatch.SingleConsumerOnlyUnboundedMailbox         12     50  thrpt   14  41823334.411 ±  753856.823  ops/s

jdk 8, fork-join-executor-akka
[info] ForkJoinActorBenchmark.pingPongLessActorsThanCores  akka.dispatch.SingleConsumerOnlyUnboundedMailbox         12     50  thrpt   14   9146008.496 ± 441612.274  ops/s
[info] ForkJoinActorBenchmark.pingPongMoreActorsThanCores  akka.dispatch.SingleConsumerOnlyUnboundedMailbox         12     50  thrpt   14  22275399.544 ± 389903.541  ops/s

jdk 11, jdk fjp
[info] ForkJoinActorBenchmark.pingPongLessActorsThanCores  akka.dispatch.SingleConsumerOnlyUnboundedMailbox         12     50  thrpt   14  12145581,861 ±  329787,280  ops/s
[info] ForkJoinActorBenchmark.pingPongMoreActorsThanCores  akka.dispatch.SingleConsumerOnlyUnboundedMailbox         12     50  thrpt   14  62261995,699 ± 3172988,446  ops/s

jdk 11, jdk fjp - second run
[info] ForkJoinActorBenchmark.pingPongLessActorsThanCores  akka.dispatch.SingleConsumerOnlyUnboundedMailbox         12     50  thrpt   14  13093229,361 ±  207326,649  ops/s
[info] ForkJoinActorBenchmark.pingPongMoreActorsThanCores  akka.dispatch.SingleConsumerOnlyUnboundedMailbox         12     50  thrpt   14  64153427,210 ± 1237483,998  ops/s

jdk 11, fork-join-executor-akka
[info] ForkJoinActorBenchmark.pingPongLessActorsThanCores  akka.dispatch.SingleConsumerOnlyUnboundedMailbox         12     50  thrpt   14   8318796,377 ±  405307,546  ops/s
[info] ForkJoinActorBenchmark.pingPongMoreActorsThanCores  akka.dispatch.SingleConsumerOnlyUnboundedMailbox         12     50  thrpt   14  19453880,183 ± 3400044,575  ops/s
```

I also tried `akka.actor.ActorBenchmark` and that is also better with JDK 11. 70 M msg/s vs 60 M msg/s.

Here is the cpu history for the ForkJoinActorBenchmark. JDK FJP at the left and akka FJP right.

Screenshot 2019-05-08 at 19 12 39

File: akka-actor/src/main/mima-filters/2.5.x.backwards.excludes [hunk 1]
```diff
@@ -21,3 +21,24 @@ ProblemFilters.exclude[MissingClassProblem]("akka.actor.dungeon.UndefinedUidActo
 
 # Protect internals against starvation #23576
 ProblemFilters.exclude[DirectMissingMethodProblem]("akka.dispatch.Dispatchers.this")
+
+# Remove internal Akka fork of FJP, was meant-to-be internal
+ProblemFilters.exclude[Problem]("akka.dispatch.forkjoin.*")
+
+# Consequences of removing the internal FJP is that class hierarchies changes. Many of those APIs are internal, others are not
+# but probably not meant to be used standalone so it's probably ok. The changes would probably only be observable if you
+# also referenced classes from akka.dispatch.forkjoin
```
Comment (patriknw) [inline akka-actor/src/main/mima-filters/2.5.x.backwards.excludes]: yes, I can't see much risk with those. 

Maybe we could look closer at what happens if a third-party library is `MonitorableThreadFactory.newThread` via `ExtendedActorSystem.threadFactory.newThread`. If that library was built with Akka 2.5.22 can it still run with Akka 2.6 (this change)?

If that is a problem we could just mention it in migration guide.

Comment (jrudolph) [reply akka-actor/src/main/mima-filters/2.5.x.backwards.excludes]: I tried it out but couldn't find ways to crash it. `threadFactory` returns a `ThreadFactory` and `newThread` a `Thread`, so it's hard to see how internal-ish types could escape there without casting these to internal things?

Comment (patriknw) [reply akka-actor/src/main/mima-filters/2.5.x.backwards.excludes]: excellent, thanks for trying

**Review by patriknw — APPROVED** [review]
LGTM

**Comment by patriknw** [comment]
Can we wait with merging this until tomorrow (Friday) so I get one more run of the tcp LatencySpec to verify that the internal-dispatcher changes actually improved things.

**Comment by jrudolph** [comment]
> Can we wait with merging this until tomorrow (Friday) so I get one more run of the tcp LatencySpec to verify that the internal-dispatcher changes actually improved things.

Sounds good!

**Review by chbatey — APPROVED** [review]
LGTM

File: akka-actor/src/main/mima-filters/2.5.x.backwards.excludes [hunk 2]
```diff
@@ -21,3 +21,24 @@ ProblemFilters.exclude[MissingClassProblem]("akka.actor.dungeon.UndefinedUidActo
...
+ProblemFilters.exclude[IncompatibleMethTypeProblem]("akka.dispatch.ForkJoinExecutorConfigurator#ForkJoinExecutorServiceFactory.this")
+ProblemFilters.exclude[IncompatibleMethTypeProblem]("akka.dispatch.ForkJoinExecutorConfigurator#ForkJoinExecutorServiceFactory.this")
+ProblemFilters.exclude[IncompatibleResultTypeProblem]("akka.dispatch.ForkJoinExecutorConfigurator.validate")
+ProblemFilters.exclude[MissingTypesProblem]("akka.dispatch.ForkJoinExecutorConfigurator$AkkaForkJoinTask")
+ProblemFilters.exclude[MissingTypesProblem]("akka.dispatch.MonitorableThreadFactory")
+ProblemFilters.exclude[IncompatibleMethTypeProblem]("akka.dispatch.MonitorableThreadFactory.newThread")
+ProblemFilters.exclude[MissingTypesProblem]("akka.dispatch.ForkJoinExecutorConfigurator$AkkaForkJoinPool")
+ProblemFilters.exclude[IncompatibleMethTypeProblem]("akka.dispatch.ForkJoinExecutorConfigurator#AkkaForkJoinPool.this")
+ProblemFilters.exclude[IncompatibleMethTypeProblem]("akka.dispatch.ForkJoinExecutorConfigurator#AkkaForkJoinPool.this")
+ProblemFilters.exclude[MissingTypesProblem]("akka.dispatch.Mailbox")
```
[the remark this answers is not in the document]
Comment (jrudolph) [reply akka-actor/src/main/mima-filters/2.5.x.backwards.excludes]: It seems a new class hierarchy counts as `MissingTypesProblem`.

**Pull request merged**